### PR TITLE
Correct set accounting accounts for auto generated invoices on multicompany environment

### PR DESCRIPTION
**Pull request by fedexin40**
On multicompany environment the create recurring invoices function will create all the invoices with accounting accounts belong to company currently selected by admin user.

This merge request solve that situation.

### CODE REVIEW

File: account_analytic_analysis_recurring/account_analytic_analysis_recurring.py [hunk 1]
```diff
@@ -241,8 +241,30 @@ def recurring_create_invoice(self, cr, uid, automatic=False, context=None):
...
             context['old_date'] = old_date
             context['next_date'] = new_date
-            self._prepare_invoice(
-                cr, uid, contract, context=context)
+            # Force company for correct evaluate domain access rules
+            context['force_company'] = contract.company_id.id
+
+            # Create new cursor for handle multi company environments
+            from openerp import pooler
+            db, pool = pooler.get_db_and_pool(cr.dbname)
```
Comment (pedrobaeza) [inline account_analytic_analysis_recurring/account_analytic_analysis_recurring.py]: You don't need a new cursor for this purpose.

**Comment by agb80** [comment]
Hello @pedrobaeza , I'm @fedexin40  co worker and in fact the one to propose the solution. Please advice us on the best way to work on this scenario:

The customer is an ISP provider with 3 companies for sell same services to different customers, we are using contracts and this module to auto generate invoices every month but the problem is as follow:
- Companies A, B & C generate contracts on the system

Case 1: Contract belong to  company A & admin user currently logged in on Company A
- Invoice Company: Company A (Correct)
- Invoice Account -> belongs to company A (correct)
- Product account -> belongs to company A (correct)

Case 2: Contract belong to  company B & admin user currently logged in on Company A
- Invoice Company: Company B (Correct)
- Invoice Account -> belongs to company A (wrong)
- Product account -> belongs to company A (wrong)

Case 3: Contract belong to  company C & admin user currently logged in on Company A
- Invoice Company: Company C (Correct)
- Invoice Account -> belongs to company A (wrong)
- Product account -> belongs to company A (wrong)

Make several test and observed that Invoice & Product account where selected per the company where admin user where currently logged in, that's why we tried the fix including only the  `context['force_company'] = contract.company_id.id`` before calling the _prepare_invoice function.

Then we face another problem, when calling the _prepare_invoice function first time, ORM optimizations where populating all the partner objects at same time resulting in a similar behavior than before:

Case 1: Contract belong to  company A & first invoice belongs to Company A
- Invoice Company: Company A (Correct)
- Invoice Account -> belongs to company A (correct)
- Product account -> belongs to company A (correct)

Case 2: Contract belong to  company B & first invoice belongs to Company A
- Invoice Company: Company B (Correct)
- Invoice Account -> belongs to company A (wrong)
- Product account -> belongs to company A (wrong)

Case 3: Contract belong to  company C & first invoice belongs to Company A
- Invoice Company: Company C (Correct)
- Invoice Account -> belongs to company A (wrong)
- Product account -> belongs to company A (wrong)

That's why we decide to create a new cursor that forces ORM populate the information for the current contract only, is working right in production at this moment but would like to learn if there is a better way to implement this.

**Comment by pedrobaeza** [comment]
Hi, Agustín,

Thanks for the huge explanation, and I understand your motivations, but I don't know if you have tried before to make a browse for each contract after setting context with force_company key. As I see in your code, you are using same browse, and that's why you receive wrong data. I think this code can work:

```
for contract in self.browse(cr, uid, contract_ids, context=context):
    # Force company to evaluate correctly domain access rules
    context['force_company'] = contract.company_id.id
    # Re-read contract with correct company
    contract = self.browse(cr, uid, contract.id, context=context)
    ...
```

**Comment by agb80** [comment]
Hello again Pedro, no we didn't have try that approach, let us check and in case needed correct back the merge request.

**Comment by agb80** [comment]
Just notice that @fedexin40 forget notify that the commit with changes asked for @pedrobaeza are live now. Please review again, thanks a lot!

**Comment by pedrobaeza** [comment]
Thanks for the changes.

:+1:

**Comment by dreispt** [comment]
It would be nice to have a test, but :+1:

**Pull request merged**